torchvision/models/mobilevit.py

```diff
@@ -0,0 +1,451 @@
+        """
+        Unfold a batch of B image tensors B x d x H X W into a batch of B  P x N x d tensors
+        (N is the number of patches)
+        These P x N x d tensors are then used by the transformer encoder where d is the hidden
+        dimension/encoding, N is the sequence length and we loop over the pixels P.
+        """
+        h_patch, w_patch = patch_size
+        n_h_patch, n_w_patch = n_patches
+        # P is the number of pixels
+        P = h_patch * w_patch
+        B, d, _, _ = x.shape
+        N = n_w_patch * n_h_patch
+
+        # We reshape from B x d x H x W to (B * d * n_h_patch) x h_patch x n_w_patch x w_patch
+        x = x.reshape(B * d * n_w_patch, h_patch, n_h_patch, w_patch)
+        # Then we transpose  (B * d * n_h_patch) x h_patch x n_w_patch x w_patch into (B * d * n_h_patch) x  n_w_patch x h_patch x w_patch
+        x = x.transpose(1, 2)
+        # Next, we reshape (B * d * n_h_patch) x  n_w_patch x h_patch x w_patch into B x d x N x P
+        x = x.reshape(B, d, N, P)
+        # And we finish by transposing B x d x N x P into B x P x N x d
+        x = x.transpose(1, 3)
+        return x
+
+    @staticmethod
+    def _fold(x: Tensor, patch_size: Tuple[int, int], n_patches: Tuple[int, int]) -> Tensor:
+        """
+        Fold a batch of B  P x N x d tensors
+        (N is the number of patches) into a batch of B d x H x W image tensors.
+        This is the reverse operation of unfold.
+        """
+        h_patch, w_patch = patch_size
+        n_h_patch, n_w_patch = n_patches
+        B, _, _, d = x.shape
+        x = x.transpose(1, 3)
+
+        x = x.reshape(B * d * n_h_patch, n_w_patch, h_patch, w_patch)
+        x = x.transpose(1, 2)
+        x = x.reshape(B, d, n_h_patch * h_patch, n_w_patch * w_patch)
+        return x
+
+    def forward(self, x):
+        # We compute how many patches along the width patch dimension, the height patch dimension,
+        # and the total number of patches.
+        # The number of patches N x the numbre of pixels P in a patch
+        # is equal to the image area H x W.
+        _, _, H, W = x.shape
+        h_patch, w_patch = self.patch_size
+        n_w_patch = W // w_patch
+        n_h_patch = H // h_patch
+        n_patches = (n_h_patch, n_w_patch)
+        y = x.detach().clone()
+        x = self.conv1(x)
+        x = self.conv2(x)
+        x = self._unfold(x, patch_size=self.patch_size, n_patches=n_patches)
+        # We get a tensor of shape: B x P x N x d after the previous steps
+        x = self.transformer(x)
+        # The transformer blocks keep the B x P x N x d shape
+        x = self._fold(x, patch_size=self.patch_size, n_patches=n_patches)
+        # We get back B x d x H x W tensors
+        x = self.conv3(x)
+        # Then we get the inital shape B x C x H X W
+        x = torch.cat((x, y), 1)
+        x = self.conv4(x)
+        return x
+
+
+# TODO: Is this separable self-attention?
+# TODO: Is this necessary? Probably for the V2 version...
```

Comment:
Shall we temporarily remove all TODO `V2` references to reduce the surface of the PR?

Reply:
I think it is a good idea indeed. I will make another PR for V2. 👍